packages/@react-stately/grid/src/useGridState.ts

```diff
@@ -46,17 +46,56 @@ export function useGridState<T extends object, C extends GridCollection<T>>(prop
     setFocusedKey(key, child);
   };
 
+  let selectionManager = useMemo(() =>
+    new SelectionManager(collection, selectionState)
+    , [collection, selectionState]
+  );
+
   // Reset focused key if that item is deleted from the collection.
+  const cachedCollection = useRef(null);
   useEffect(() => {
     if (selectionState.focusedKey != null && !collection.getItem(selectionState.focusedKey)) {
-      selectionState.setFocusedKey(null);
+      const node = cachedCollection.current.getItem(selectionState.focusedKey);
+      const parentNode =
+        node.parentKey != null && (node.type === 'cell' || node.type === 'rowheader' || node.type === 'column') ?
+        cachedCollection.current.getItem(node.parentKey) :
+        node;
+      const rows = collection.rows;
+      let index = Math.min(parentNode.index, rows.length - 1);
+      let newRow:GridNode<T>;
+      while (index >= 0) {
+        if (!selectionManager.isDisabled(rows[index].key)) {
+          newRow = rows[index];
+          break;
+        }
+        // Find next, not disabled row.
+        if (index < rows.length - 1) {
+          index++;
+        // Otherwise, find previous, not disabled row.
+        } else {
+          if (index > parentNode.index) {
+            index = parentNode.index;
+          }
+          index--;
+        }
+      }
+      if (newRow) {
+        const keyToFocus =
+          newRow.hasChildNodes && parentNode !== node ?
+          [...newRow.childNodes][node.index].key :
+          newRow.key;
```

Comment:
What if the grid's column is deleted? I _think_ this might run into a problem since `node.index` won't exist in the new row's childNodes

Reply:
Would this work to ensure that the `node.index` is present in the `newRow`?
```suggestion
        const keyToFocus =
          newRow.hasChildNodes  && parentNode !== node && node.index < newRow.childNodes.length  ?
          [...newRow.childNodes][node.index].key :
          newRow.key;
```

Reply:
At a glance I think so? I can try experimenting with a table whose columns/contents change after a settimeout

Reply:
Cool, verified this locally, seems to work well